main/src/library/Set.flix

```diff
@@ -324,9 +324,15 @@ namespace Set {
     ///
     /// Returns the union of `xs` and `ys`.
     ///
-    @Time(size(ys) * Int32.log2(size(xs))) @Space(size(ys) * Int32.log2(size(xs)))
+    @Time(Int32.min(size(xs), size(ys)) * Int32.log2(size(xs) + size(ys)))
+    @Space(Int32.min(size(xs), size(ys)) * Int32.log2(size(xs) + size(ys)))
     pub def union[a : Order](xs: Set[a], ys: Set[a]): Set[a] =
-        foldLeft((acc, x) -> insert(x, acc), xs, ys)
+        let Set(as) = xs;
+        let Set(bs) = ys;
+        if (RedBlackTree.height(as) < RedBlackTree.height(bs))
+            foldLeft((acc, x) -> insert(x, acc), ys, xs)
```

Comment:
By using xs and ys here, I think you are spending an extra allocation of `Set` per size of the smaller set. This could possibly also be avoided?

Reply:
Could also say that this should be fixed by having opaque types actually being erased, so it is okay if you want to leave it.

Reply:
For now, let's just say that everything counts :)